psst-gui/src/webapi/client.rs

```diff
@@ -821,33 +824,38 @@ impl WebApi {
     }
 
     pub fn get_lyrics(&self, track_id: String) -> Result<Vector<TrackLines>, Error> {
-        #[derive(Default, Debug, Clone, PartialEq, Deserialize, Data)]
-        #[serde(rename_all = "camelCase")]
-        pub struct Root {
-            pub lyrics: Lyrics,
+        let track_id_arc: Arc<str> = track_id.clone().into();
+
+        // Check in-memory cache
+        if let Some(cached_lyrics) = self.cache.get_lyrics(&track_id_arc) {
+            return Ok(cached_lyrics);
         }
 
-        #[derive(Default, Debug, Clone, PartialEq, Deserialize, Data)]
-        #[serde(rename_all = "camelCase")]
-        pub struct Lyrics {
-            pub lines: Vector<TrackLines>,
-            pub provider: String,
-            pub provider_lyrics_id: String,
+        // Check disk cache
+        if let Some(disk_cached_lyrics) = self.cache.get_lyrics_from_disk(&track_id_arc) {
+            self.cache
+                .set_lyrics(track_id_arc.clone(), disk_cached_lyrics.clone());
+            return Ok(disk_cached_lyrics);
         }
 
+        // If not in cache, fetch from API
         let token = self.access_token()?;
         let request = self
             .agent
             .request("GET", &format!("https://spclient.wg.spotify.com/color-lyrics/v2/track/{}/image/https%3A%2F%2Fi.scdn.co%2Fimage%2F{}", track_id, track_id.clone().split_off(3)))
             .query("format", "json")
             .query("vocalRemoval", "false")
             .query("market", "from_token")
-            // This is the reason for redoing the request method here rather than reusing as the order needs to be this way round.
             .set("app-platform", "WebPlayer")
             .set("Authorization", &format!("Bearer {}", &token));
 
-        let result: Cached<Root> = self.load_cached(request.clone(), "Lyrics", &track_id)?;
-        Ok(result.data.lyrics.lines)
+        let lyrics: Vector<TrackLines> = self.load(request)?;
+
+        // Save to in-memory and disk cache
+        self.cache.set_lyrics(track_id_arc.clone(), lyrics.clone());
+        self.cache.save_lyrics_to_disk(&track_id_arc, &lyrics);
+
+        Ok(lyrics)
```

Comment:
Was your method of caching saving it to disk as well? Is this excessive?

Reply:
Oh are you caching the request? Maybe that's a better way to do this. I generally think we need a caching overhaul where we cache all requests in a singular efficient method.

Reply:
Yes i was caching the request and the response. The way that the load_cached works is if its not there it creates a folder and i name it to the song id. I think what you have done is very similar to that function so it may be excessive. Or maybe we just keep it with load_cached to keep it the same as the other functions? Then we change the load_cached function? 

I was thinking with the caching we could include a timestamp for the chached item which then deletes the cache after a certain time otherwise it may take up a lot of space. It also could then be a setting in the settings page as an "Update time" section so the user can set how often they want it to refetch everything. 

I was also thinking that using lazy loading we could cache just say the first page (50) or so (perhaps the user could define this) which would mean that we could show it very quickly to the user then the rest is brought in when the user is at the bottom of the page. (Or perhaps at the start we could have a show more button?)

We would have to log two things, user position in the list and then also the offset needed, this would also then mean that we could save the place the user is on the playlist which would be a very nice addition.

Reply:
Also, you removed the Deserialization, which has made it so it doesn't work on my machine. Did it work on yours with out it?

Reply:
Oh I think we should use this load_cached as the go to function. Sorry for complicating it, we can eventually integrate that method into every call.